File: worker/src/lib/pagination.ts
export const DEFAULT_PAGE_LIMIT = 50;
export const MAX_PAGE_LIMIT = 500;

export type Cursor = {
  afterIntID?: number;
  afterStringID?: string;
  afterCreated?: string;
};

function encodeB64UrlString(s: string): string {
  return btoa(s).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

function decodeB64UrlString(s: string): string {
  let b64 = s.replace(/-/g, '+').replace(/_/g, '/');
  while (b64.length % 4) b64 += '=';
  return atob(b64);
}

export function encodeCursor(c: Cursor): string {
  const obj: Record<string, unknown> = {};
  if (c.afterIntID !== undefined && c.afterIntID !== 0) obj['i'] = c.afterIntID;
  if (c.afterStringID) obj['s'] = c.afterStringID;
  if (c.afterCreated) obj['t'] = c.afterCreated;
  return encodeB64UrlString(JSON.stringify(obj));
}

export function decodeCursor(s: string | null | undefined): Cursor {
  if (!s) return {};
  let raw: string;
  try {
    raw = decodeB64UrlString(s);
  } catch {
    throw new Error('invalid cursor');
  }
  let obj: Record<string, unknown>;
  try {
    obj = JSON.parse(raw);
  } catch {
    throw new Error('invalid cursor');
  }
  const c: Cursor = {};
  if (typeof obj['i'] === 'number') c.afterIntID = obj['i'];
  if (typeof obj['s'] === 'string') c.afterStringID = obj['s'];
  if (typeof obj['t'] === 'string') c.afterCreated = obj['t'];
  return c;
}

export function clampLimit(n: number | null | undefined): number {
  if (!n || n <= 0) return DEFAULT_PAGE_LIMIT;
  if (n > MAX_PAGE_LIMIT) return MAX_PAGE_LIMIT;
  return Math.floor(n);
}

File: src/internal/store/pagination.go
package store

import (
	"encoding/base64"
	"encoding/json"
	"errors"
)

const (
	DefaultPageLimit = 50
	MaxPageLimit     = 500
)

type Cursor struct {
	AfterIntID    int64  `json:"i,omitempty"`
	AfterStringID string `json:"s,omitempty"`
	AfterCreated  string `json:"t,omitempty"`
}

func (c Cursor) Encode() string {
	b, _ := json.Marshal(c)
	return base64.RawURLEncoding.EncodeToString(b)
}

func DecodeCursor(s string) (Cursor, error) {
	if s == "" {
		return Cursor{}, nil
	}
	raw, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		return Cursor{}, errors.New("invalid cursor")
	}
	var c Cursor
	if err := json.Unmarshal(raw, &c); err != nil {
		return Cursor{}, errors.New("invalid cursor")
	}
	return c, nil
}

func ClampLimit(n int) int {
	if n <= 0 {
		return DefaultPageLimit
	}
	if n > MaxPageLimit {
		return MaxPageLimit
	}
	return n
}
